### PR TITLE
[BULK] Shard id population in transformer

### PR DESCRIPTION
**Pull request by Deep1998**

### CODE REVIEW

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/options/SourceDbToSpannerOptions.java [hunk 1]
```diff
@@ -176,4 +176,16 @@ public interface SourceDbToSpannerOptions extends CommonTemplateOptions {
...
+
+  @TemplateParameter.GcsReadFile(
+      order = 15,
+      optional = true,
+      helpText =
+          "Transformation context file path in cloud storage used to populate data used in"
+              + " transformations performed during migrations. Eg: Maps db to shard id to"
+              + " identify the db from which shard a row was migrated",
+      description = "Transformation context file path in cloud storage")
+  String getTransformationContextFilePath();
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/options/SourceDbToSpannerOptions.java:188]: I'm assuming this is the shard configuration file. We should name it the same.

Comment (Deep1998) [reply v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/options/SourceDbToSpannerOptions.java:188]: shardConfig is used in reverse replication. For forward, we use this only.

**Review by VardhanThigle — APPROVED** [review]

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/schema/SessionBasedMapper.java [hunk 2]
```diff
@@ -149,4 +155,22 @@ public List<String> getSpannerColumns(String namespace, String spannerTable)
...
+  public String getShardIdColumnName(String namespace, String spannerTableName) {
+    Map<String, NameAndCols> spanToId = schema.getSpannerToID();
+    if (!spanToId.containsKey(spannerTableName)) {
+      throw new NoSuchElementException(
+          String.format("Spanner table '%s' not found", spannerTableName));
+    }
+    try {
+      String tableId = spanToId.get(spannerTableName).getName();
+      SpannerTable table = schema.getSpSchema().get(tableId);
+      String colId = table.getShardIdColumn();
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/schema/SessionBasedMapper.java:169]: If colId is null, or does not exist in col definition, we should throw an explicit exception. Catching Null pointers and throwing a runtime exception is not really a good idea. Lets used types exceptions here.

**Review by bharadwaj-aditya — CHANGES_REQUESTED** [review]

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/GenericRecordTypeConvertor.java [hunk 3]
```diff
@@ -100,6 +118,21 @@ public Map<String, Value> transformChangeEvent(GenericRecord record, String srcT
     return result;
   }
 
+  private Map<String, Value> populateShardId(Map<String, Value> result, String shardIdCol) {
+    if (transformationContext.getSchemaToShardId() == null
```
[the remark this answers is not in the document]
Comment (Deep1998) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/GenericRecordTypeConvertor.java]: Updated

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SourceDbToSpanner.java [hunk 4]
```diff
@@ -150,7 +157,11 @@ static PipelineResult run(SourceDbToSpannerOptions options) {
 
     // Transform source data to Spanner Compatible Data
     SourceRowToMutationDoFn transformDoFn =
-        SourceRowToMutationDoFn.create(schemaMapper, getTableIDToRefMap(srcSchema));
+        SourceRowToMutationDoFn.create(
+            schemaMapper,
+            getTableIDToRefMap(srcSchema),
+            transformationContext,
```
[the remark this answers is not in the document]
Comment (Deep1998) [reply v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SourceDbToSpanner.java]: Removed in the latest iupdate.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/schema/SessionBasedMapper.java (hunk 2 above)
Comment (Deep1998) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/schema/SessionBasedMapper.java]: Updated

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM

**Pull request merged**